File: cannabis_management/credit_and_ar/web_form/credit_agreement/credit_agreement.js
frappe.ready(function () {
	// bind events here

	// Show the company name, centered, above the form title — mirrors the
	// boxed "[COMPANY NAME]" pill at the top of the Credit Agreement PDF.
	var $head = $(".web-form-head").first();
	if ($head.length && !$head.find(".ca-companyname").length) {
		$("<div>", {
			text: "MOTLEY TERPZ",
			class: "ca-companyname",
		}).prependTo($head);
	}

	// Pre-fill today's date on a new agreement. The web form's own
	// "default": "Today" mechanism (unlike Desk) never special-cases the
	// word "Today" — it stuffs that literal string into the field, which
	// then fails date validation and pops a "Date Today must be in format:
	// ..." error. Setting the real ISO value through the control here
	// avoids that entirely.
	if (frappe.web_form && frappe.web_form.is_new && frappe.web_form.fields_dict.agreement_date) {
		frappe.web_form.set_value("agreement_date", frappe.datetime.get_today());
	}

	// Sales Rep and Payment Terms are excluded/filtered server-side, in
	// credit_agreement.py's get_context — NOT here. Web forms turn every
	// Link field into an "Autocomplete" backed by a static, pre-baked
	// option list (see WebForm.load_form_data / get_link_options); setting
	// df.get_query on that control makes it call a server *method* instead
	// (df.get_query = {query, params}, not Desk's {filters: {...}}), so a
	// Link-style filters object here just empties the list — it doesn't
	// narrow it. Filtering the baked-in options string before render is
	// the only thing that actually works for this fieldtype.

	// Match the printed agreement's "Label:" style. This has to be real text
	// appended after the label, not a CSS ::after — Frappe already puts the
	// red required-asterisk on .control-label via ::after (controls.scss),
	// and only one ::after can win per element.
	$(".web-form-wrapper .control-label").each(function () {
		var $label = $(this);
		if (!/:\s*$/.test($label.text())) {
			$label.append(":");
		}
	});

	// The PDF attached to the record on submit is rendered server-side from
	// the "Credit Agreement" Print Format (see
	// cannabis_management.credit_and_ar.web_form_intake.after_insert) — a
	// clean copy of the printed agreement filled with the values the client
	// just submitted, not a screenshot of this web page. Nothing to do here.
});

File: cannabis_management/credit_and_ar/web_form/credit_agreement/credit_agreement.py
import frappe

# Plain NET terms before their "50% down" variants, each group sorted by day
# count — the order the client asked the Payment Terms dropdown to list in.
APPROVED_TERMS_ORDER = [
	"NET7",
	"NET15",
	"NET21",
	"NET30",
	"50% down NET7",
	"50% down NET15",
	"50% down NET21",
	"50% down NET30",
]

# Never offered on this form, regardless of what's in Payment Terms Template.
APPROVED_TERMS_EXCLUDE = {"COD", "PAYMENT SCHEDULE BREAKDOWN"}


def get_context(context):
	"""Context for the public Credit Agreement web form."""
	context.no_cache = 1
	# Keep this client-facing page's title clean — skip the sitewide "MT -" prefix.
	context.title_prefix = None

	# Web forms turn every Link field into an "Autocomplete" backed by a
	# static list of every record name (see WebForm.load_form_data ->
	# get_link_options, which fetches with no filters or ordering at all)
	# baked into context.web_form_doc.web_form_fields before this hook
	# runs. There's no per-field filter/order hook for that, so the only
	# way to control what shows (and in what order) is to rewrite the
	# already-built options string here.
	_set_link_options(context, "sales_person", _sales_person_rep_names())
	_set_link_options(context, "approved_terms", _approved_terms_names())


def _sales_person_rep_names():
	"""Actual Sales Person reps — "group" org-tree nodes excluded."""
	return frappe.get_all("Sales Person", filters={"is_group": 0}, pluck="name")


def _approved_terms_names():
	existing = set(frappe.get_all("Payment Terms Template", pluck="name")) - APPROVED_TERMS_EXCLUDE
	ordered = [name for name in APPROVED_TERMS_ORDER if name in existing]
	# Any term added later that isn't in our explicit order (and isn't
	# excluded above) still shows up, just tacked on at the end, instead of
	# silently vanishing from the list.
	ordered += sorted(existing - set(ordered))
	return ordered


def _set_link_options(context, fieldname, values):
	for field in context.web_form_doc.web_form_fields:
		if field.get("fieldname") == fieldname:
			field.options = "\n".join(values)
			break
